pkg/agent/nodeportlocal/k8s/controller.go

```diff
@@ -112,12 +175,136 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+		return nil, fmt.Errorf("Received unexpected object: %v", obj)
+	}
+	svc, ok := deletedState.Obj.(*corev1.Service)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Service: %v", deletedState.Obj)
+	}
+	return svc, nil
+}
+
+func (c *NPLController) enqueueSvcUpdate(oldObj, newObj interface{}) {
+	// In case where the app selector in Service gets updated from one valid selector to another
+	// both sets of Pods (corresponding to old and new selector) need to be considered.
+	newSvc := newObj.(*corev1.Service)
+	oldSvc := oldObj.(*corev1.Service)
+
+	// Return if Service ResourceVersions don't change.
+	if oldSvc.ResourceVersion == newSvc.ResourceVersion {
+		return
+	}
+
+	oldSvcAnnotation := oldSvc.Annotations[NPLEnabledAnnotationKey]
+	newSvcAnnotation := newSvc.Annotations[NPLEnabledAnnotationKey]
+	// Return if both Services donot have the NPL annotation.
+	if oldSvcAnnotation != "true" && newSvcAnnotation != "true" {
+		return
+	}
+
+	podKeys := sets.String{}
+	if oldSvcAnnotation != newSvcAnnotation {
+		// Process Pods corresponding to Service with valid NPL annotation.
+		if oldSvcAnnotation == "true" {
+			podKeys = sets.NewString(c.getPodsFromService(oldSvc)...)
+		} else if newSvcAnnotation == "true" {
+			podKeys = sets.NewString(c.getPodsFromService(newSvc)...)
+		}
+	} else if !reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector) {
+		// Disjunctive union of Pods from both Service sets.
+		oldPodSet := sets.NewString(c.getPodsFromService(oldSvc)...)
+		newPodSet := sets.NewString(c.getPodsFromService(newSvc)...)
+		podKeys = oldPodSet.Difference(newPodSet).Union(newPodSet.Difference(oldPodSet))
+	}
+
+	for podKey := range podKeys {
+		c.queue.Add(podKey)
+	}
+}
+
+func (c *NPLController) enqueueSvc(obj interface{}) {
+	svc, isSvc := obj.(*corev1.Service)
+	if !isSvc {
+		var err error
+		svc, err = c.checkDeletedSvc(obj)
+		if err != nil {
+			klog.Errorf("Got error while processing event update: %v", err)
+			return
+		}
+	}
+
+	// Process Pods corresponding to Service with valid NPL annotation.
+	if svc.Annotations[NPLEnabledAnnotationKey] == "true" {
+		for _, podKey := range c.getPodsFromService(svc) {
+			c.queue.Add(podKey)
+		}
+	}
+}
+
+func (c *NPLController) getPodsFromService(svc *corev1.Service) []string {
+	var pods []string
+
+	// Handling Service without selectors.
+	if len(svc.Spec.Selector) == 0 {
+		return pods
+	}
+
+	podList, err := c.podLister.List(labels.SelectorFromSet(labels.Set(svc.Spec.Selector)))
+	if err != nil {
+		klog.Errorf("Got error while listing Pods: %v", err)
+		return pods
+	}
+	for _, pod := range podList {
+		pods = append(pods, podKeyFunc(pod))
+	}
+	return pods
+}
+
+func (c *NPLController) isNPLEnabledForServiceOfPod(obj interface{}) bool {
+	pod := obj.(*corev1.Pod)
+	services, err := c.svcInformer.GetIndexer().ByIndex(NPLEnabledAnnotationIndex, "true")
+	if err != nil {
+		klog.Errorf("Got error while listing Services with annotation: %v", err)
```

Comment:
maybe indicate which annotation in the log message, or that could be confusing when inspecting logs